drivers/adc/ad7616/ad7616.c

```diff
@@ -952,6 +956,18 @@ int32_t ad7616_setup(struct ad7616_dev **device,
 	}
 
 #ifdef XILINX_PLATFORM
+	ret = axi_clkgen_init(&dev->clkgen, init_param->clkgen_init);
+	if (ret != 0)
+		return ret;
+
+	ret = axi_clkgen_set_rate(dev->clkgen, init_param->axi_clkgen_rate);
+	if (ret != 0)
+		return ret;
+
+	ret = no_os_pwm_init(&dev->trigger_pwm_desc, init_param->trigger_pwm_init);
+	if (ret != 0)
+		return ret;
```

Comment:
dev should be freed before returning and axi_clkgen_remove should be called.

Reply:
also, no_os_pwm_remove should be called extra if an error occurs in the lines below.
I propose to refactor this function to treat all the potential memory leaks.